certbot/ocsp.py

```diff
@@ -65,33 +88,126 @@ def ocsp_revoked(self, cert_path, chain_path):
+        return None, None
+
+    url = url.rstrip()
+    host = url.partition("://")[2].rstrip("/")
+
+    if host:
+        return url, host
+    else:
+        logger.info("Cannot process OCSP host from URL (%s) in cert at %s", url, cert_path)
+        return None, None
+
+
+def _check_ocsp_cryptography(cert_path, chain_path, url):
+    # type: (str, str, str) -> bool
+    # Retrieve OCSP response
+    with open(chain_path, 'rb') as file_handler:
+        issuer = x509.load_pem_x509_certificate(file_handler.read(), default_backend())
+    with open(cert_path, 'rb') as file_handler:
+        cert = x509.load_pem_x509_certificate(file_handler.read(), default_backend())
+    builder = ocsp.OCSPRequestBuilder()
+    builder = builder.add_certificate(cert, issuer, hashes.SHA1())
+    request = builder.build()
+    request_binary = request.public_bytes(serialization.Encoding.DER)
+    response = requests.post(url, data=request_binary,
+                             headers={'Content-Type': 'application/ocsp-request'})
+    if response.status_code != 200:
+        logger.info("OCSP check failed for %s (are we offline?)", cert_path)
+        return False
+    response_ocsp = ocsp.load_der_ocsp_response(response.content)
+
+    # Check OCSP response validity
+    if response_ocsp.response_status != ocsp.OCSPResponseStatus.SUCCESSFUL:
+        logger.error("Invalid OCSP response status for %s: %s",
+                     cert_path, response_ocsp.response_status)
+        return False
+
+    # Check OCSP signature
+    try:
+        _check_ocsp_response(response_ocsp, request, issuer)
+    except UnsupportedAlgorithm as e:
+        logger.error(str(e))
+    except errors.Error as e:
+        logger.error(str(e))
+    except InvalidSignature:
+        logger.error('Invalid signature on OCSP response for %s', cert_path)
+    except AssertionError as error:
+        logger.error('Invalid OCSP response for %s: %s.', cert_path, str(error))
+    else:
+        # Check OCSP certificate status
+        logger.debug("OCSP certificate status for %s is: %s",
+                     cert_path, response_ocsp.certificate_status)
+        return response_ocsp.certificate_status == ocsp.OCSPCertStatus.REVOKED
+
+    return False
+
+
+def _check_ocsp_response(response_ocsp, request_ocsp, issuer_cert):
+    """Verify that the OCSP is valid for serveral criterias"""
+    # Assert OCSP response corresponds to the certificate we are talking about
+    if response_ocsp.serial_number != request_ocsp.serial_number:
+        raise AssertionError('the certificate in response does not correspond '
+                             'to the certificate in request')
+
+    # Assert signature is valid
+    _check_ocsp_response_signature(response_ocsp, issuer_cert)
+
+    # Assert issuer in response is the expected one
+    if (response_ocsp.issuer_key_hash != request_ocsp.issuer_key_hash
+            or response_ocsp.issuer_name_hash != request_ocsp.issuer_name_hash):
```

Comment:
I fought GH to let me continue the thread at https://github.com/certbot/certbot/pull/6603#discussion_r248453185 and failed so here we go...

> I agree for issuer_key_hash. But hash_algorithm is overkill. If hashes algorithms does not correspond, there is no way that hashes can correspond.

I disagree. While I agree the chance of hash collisions in two different hash functions that are supported both on the client and server side for two different values is extremely unlikely, I don't think we get any cryptographic guarantees like that. I strongly prefer we also check `hash_algorithm` and err on the safe side.

Reply:
This is highly improbable, I would bet more confidently on the Armageddon movie scenario to happen in real life next yeqr. But it is mathematically not impossible, see https://crypto.stackexchange.com/questions/5304/can-there-be-two-hash-functions-without-common-collisions. And computationally almost free for us to check also the hash algorithm.

So I add it ^^

Reply:
Thanks. I agree with everything you said :)